### PR TITLE
Add Profiler Support for Performance Analysis

### PR DESCRIPTION
**Pull request by yhna940**
# What does this PR do?

This PR introduces profiler support to the Accelerate library, enabling users to collect performance metrics during model training and inference. The profiler allows for detailed analysis of execution time and memory consumption of model operators and can generate profiling traces for visualization in Chrome's tracing tool. Additionally, it provides options to profile long-running jobs with customizable scheduling options.

profile_export

Key changes include:
- Addition of `ProfileKwargs` to customize profiling options.
- Implementation of profiling context manager in the `Accelerate` class.
- New documentation explaining how to use the profiler with examples.
- Updated examples to demonstrate profiling capabilities (`profiler.py`).
- Tests for profiling functionality with various scheduling options.

## Context and Motivation
Other frameworks like [MMEngine](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.hooks.ProfilerHook.html) and [PyTorch Lightning](https://pytorch-lightning.readthedocs.io/en/1.5.10/advanced/profiler.html) offer profiling techniques based on the [Torch Profiler](https://pytorch.org/docs/stable/profiler.html). Inspired by these tools, we aimed to bring similar profiling capabilities to Accelerate. This enhancement helps users optimize and improve model performance by providing insights into the computational and memory aspects of their models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for implementing this cool feature. At a first glance, the PR already looks super clean, really great job.

I checked the [rendered profiler docs](https://moon-ci-docs.huggingface.co/docs/accelerate/pr_2883/en/usage_guides/profiler) and there seems to be an issue with `</hfoption>`, not sure what's going on there. Could you please check? _Edit_: It seems you were faster than me :)

I also saw that the dates of the copyright headers you added are outdated, could you please update them to 2024?

Otherwise, this PR looks great to me. I don't have a ton of experience with the profiler -- except that I found that it can be super slow :D. In terms of the design of the feature, it looks quite nice to me, but let's wait for Zach's return to office for a full review.

**Comment by yhna940** [comment]
> Thanks for implementing this cool feature. At a first glance, the PR already looks super clean, really great job.
> 
> I checked the [rendered profiler docs](https://moon-ci-docs.huggingface.co/docs/accelerate/pr_2883/en/usage_guides/profiler) and there seems to be an issue with `</hfoption>`, not sure what's going on there. Could you please check? _Edit_: It seems you were faster than me :)
> 
> I also saw that the dates of the copyright headers you added are outdated, could you please update them to 2024?
> 
> Otherwise, this PR looks great to me. I don't have a ton of experience with the profiler -- except that I found that it can be super slow :D. In terms of the design of the feature, it looks quite nice to me, but let's wait for Zach's return to office for a full review.


Thank you for the quick review @BenjaminBossan !

I appreciate you pointing out the issue with the `</hfoption>` tag in the docs. I've corrected it now. Additionally, I've updated the copyright headers to 2024 as requested.

However, I noticed that one image in the docs is broken. Could you please provide some guidance on how to fix this? 🙏

Regarding the profiler's performance, I'm also interested in investigating the slowdown you mentioned. On my local server, it seems to work fine. Could you provide more details on where you're experiencing the slowness?

Thanks again for your feedback :)

File: examples/by_feature/profiler.py
Comment (BenjaminBossan) [inline examples/by_feature/profiler.py:1]: Still old year.

File: src/accelerate/accelerator.py
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:3423]: How about adding the `_` to the `PROFILE_PATTERN_NAME` directly?

File: src/accelerate/utils/dataclasses.py
Comment (BenjaminBossan) [inline src/accelerate/utils/dataclasses.py:358]: I think for this docstring to be included in the API docs, you need to add an entry to `docs/source/package_references/kwargs.md`.

File: tests/test_examples.py
Comment (BenjaminBossan) [inline tests/test_examples.py:252]: At least on my machine, this takes a significant amount of time to run, should `@slow` be added?

File: tests/test_kwargs_handlers.py
Comment (BenjaminBossan) [inline tests/test_kwargs_handlers.py:124]: I wonder: Instead of printing, could we, for instance, just add this to a list, then do a basic sanity check on the output, for example what columns we should expect?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
This looks pretty good already, I have only some minor comments. As mentioned, the full review should come from Zach when he's back.

> However, I noticed that one image in the docs is broken.

I wonder if it would ahve been okay as it was, once the PR would be merged. Currently, the image is indeed not at the original location, but that would change after the merge. Unfortunately, I couldn't find any example in the existing docs to verify this.

> Regarding the profiler's performance, I'm also interested in investigating the slowdown you mentioned. On my local server, it seems to work fine. Could you provide more details on where you're experiencing the slowness?

From memory, when I tested it, I found some sources of slow down:

- Processing the data seems to take quite some time when the model is big.
- Some options (IIRC `with_stack`) would considerably slow down the run itself.
- The generated trace could be huge (>1GB) and bring Chrome to its knees.

Of course, this is to be expected to some extent but it made me not want to use the feature ;)

File: examples/by_feature/profiler.py
Comment (yhna940) [reply examples/by_feature/profiler.py]: Done `20f377b`

File: src/accelerate/accelerator.py
Comment (yhna940) [reply src/accelerate/accelerator.py]: Done `9cfbbf3`

File: src/accelerate/utils/dataclasses.py
Comment (yhna940) [reply src/accelerate/utils/dataclasses.py:358]: Done `06a2430`

File: tests/test_examples.py
Comment (yhna940) [reply tests/test_examples.py:252]: Thank you :) I totally agree with your comment. I added a `slow` decorator to the function.

**Comment by yhna940** [comment]
Thank you for the review and the feedback! I appreciate the insights and will wait for Zach's full review for any additional comments or suggestions.

I will conduct some experiments to further investigate the performance impacts of `with_stack` and data processing times.

Additionally, I propose updating the user guide to include tips or warnings about handling large or complex tasks using the PyTorch profiler's API for long-running jobs. As highlighted in the [PyTorch official documentation](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html):

> PyTorch profiler offers an additional API to handle long-running jobs (such as training loops). Tracing all of the execution can be slow and result in very large trace files. To avoid this, use optional arguments:
> 
> - `schedule` - specifies a function that takes an integer argument (step number) as an input and returns an action for the profiler, the best way to use this parameter is to use the `torch.profiler.schedule` helper function that can generate a schedule for you;
> - `on_trace_ready` - specifies a function that takes a reference to the profiler as an input and is called by the profiler each time the new trace is ready.

By updating the documentation with these details, we can provide more comprehensive guidance to users, helping them to effectively manage performance profiling for their models.

Let me know if there's anything else I should address or if you have further suggestions!

Thanks again for the review :)

**Review by BenjaminBossan — APPROVED** [review]
Thanks a lot for the updates. From my point of view, this looks good. As mentioned, we'll wait for Zach for the final review.

> I will conduct some experiments to further investigate the performance impacts of with_stack and data processing times.

> By updating the documentation with these details, we can provide more comprehensive guidance to users, helping them to effectively manage performance profiling for their models.

I wouldn't put too much time into this, as I'm sure this is a moving target and we can't constantly monitor the PyTorch docs for updates. I think it's sufficient to refer users to the PyTorch docs for caveats and performance tips.

**Review by muellerzr — APPROVED** [review]
This is an exceedingly clean, thorough, and excellent PR @yhna940, very nice work! Let's get this in 🔥 

cc @stas00

**Pull request merged**
